pycontrails/datalib/ecmwf/model_level_hres.py

```diff
@@ -0,0 +1,511 @@
+        LOG.debug(f"Retrieving ERA5 data for times {times} from forecast {self.forecast_time}")
+        _download_convert_cache_handler(self, times)
+
+    @overrides
+    def open_metdataset(
+        self,
+        dataset: xr.Dataset | None = None,
+        xr_kwargs: dict[str, Any] | None = None,
+        **kwargs: Any,
+    ) -> MetDataset:
+
+        if dataset:
+            msg = "Parameter 'dataset' is not supported for Model-level ERA5 data"
+            raise ValueError(msg)
+
+        if self.cachestore is None:
+            msg = "Cachestore is required to download data"
+            raise ValueError(msg)
+
+        xr_kwargs = xr_kwargs or {}
+        self.download(**xr_kwargs)
+
+        disk_cachepaths = [self.cachestore.get(f) for f in self._cachepaths]
+        ds = self.open_dataset(disk_cachepaths, **xr_kwargs)
+
+        mds = self._process_dataset(ds, **kwargs)
+
+        self.set_metadata(mds)
+        return mds
+
+    @overrides
+    def set_metadata(self, ds: xr.Dataset | MetDataset) -> None:
+        ds.attrs.update(
+            provider="ECMWF", dataset="HRES", product="forecast", radiation_accumulated=True
+        )
+
+    def mars_request(self, times: list[datetime]) -> str:
+        """Generate MARS request for specific list of times.
+
+        Parameters
+        ----------
+        times : list[datetime]
+            Times included in MARS request.
+
+        Returns
+        -------
+        str
+            MARS request for submission to ECMWF API.
+        """
+        date = self.forecast_time.strftime("%Y-%m-%d")
+        time = self.forecast_time.strftime("%H:%M:%S")
+        steps = self.get_forecast_steps(times)
+        # param 152 = log surface pressure, needed for metview level conversion
+        grib_params = set(self.variable_ecmwfids + [152])
+        return (
+            f"retrieve,\n"
+            f"class=od,\n"
+            f"date={date},\n"
+            f"expver=1,\n"
+            f"levelist={'/'.join(str(lev) for lev in sorted(self.levels))},\n"
+            f"levtype=ml,\n"
+            f"param={'/'.join(str(p) for p in sorted(grib_params))},\n"
+            f"step={'/'.join(str(s) for s in sorted(steps))},\n"
+            f"stream=oper,\n"
+            f"time={time},\n"
+            f"type=fc,\n"
+            f"grid={self.grid}/{self.grid}"
+        )
+
+    def _set_server(self) -> None:
+        """Set the ecmwfapi.ECMWFService instance."""
+        try:
+            from ecmwfapi import ECMWFService
+        except ModuleNotFoundError as e:
+            dependencies.raise_module_not_found_error(
+                name="ModelLevelHRES._set_server method",
+                package_name="ecmwf-api-client",
+                module_not_found_error=e,
+                pycontrails_optional_package="ecmwf",
+            )
+
+        self.server = ECMWFService("mars", url=self.url, key=self.key, email=self.email)
+
+
+def _download_convert_cache_handler(
```

Comment:
Similar to comment to era5 - is there a reason this is outside the class?

Reply:
Nope--will move this inside too!